src/compute/src/compute_state/peek_stash.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+//! For eligible peeks, we send the result back via the peek stash (aka persist
+//! blob), instead of inline in `ComputeResponse`.
+
+use std::num::{NonZeroI64, NonZeroU64};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use mz_compute_client::protocol::command::Peek;
+use mz_compute_client::protocol::response::{PeekResponse, StashedPeekResponse};
+use mz_expr::row::RowCollection;
+use mz_ore::cast::CastFrom;
+use mz_ore::task::AbortOnDropHandle;
+use mz_persist_client::Schemas;
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_types::codec_impls::UnitSchema;
+use mz_persist_types::{PersistLocation, ShardId};
+use mz_repr::{Diff, RelationDesc, Row, Timestamp};
+use mz_storage_types::sources::SourceData;
+use timely::progress::Antichain;
+use tokio::sync::mpsc::error::TrySendError;
+use tokio::sync::oneshot;
+use tracing::debug;
+use uuid::Uuid;
+
+use crate::arrangement::manager::{PaddedTrace, TraceBundle};
+use crate::compute_state::peek_result_iterator;
+use crate::compute_state::peek_result_iterator::PeekResultIterator;
+use crate::typedefs::RowRowAgent;
+
+/// An async task that stashes a peek response in persist and yield a handle to
+/// the batch in a [PeekResponse::Stashed].
+///
+/// Note that `PeekResponseTask` intentionally does not implement or derive
+/// `Clone`, as each `PeekResponseTask` is meant to be dropped after it's
+/// done or no longer needed.
+pub struct StashPeekResponse {
+    pub peek: Peek,
+    /// Iterator for the results. The worker thread has to continually pump
+    /// results from this to the `rows_tx` channel.
+    peek_iterator: Option<PeekResultIterator<PaddedTrace<RowRowAgent<Timestamp, Diff>>>>,
+    /// We can't give a PeekResultIterator to our async upload task because the
+    /// underlying trace reader is not Send/Sync. So we need to use a channel to
+    /// send result rows from the worker thread to the async background task.
+    rows_tx: Option<tokio::sync::mpsc::Sender<Result<Vec<(Row, NonZeroI64)>, String>>>,
+    /// When the async task is busy it cannot work of rows, and we don't want to
+    /// use an unbounded channel. We stash rows that we cannot send and retry
+    /// later.
+    stashed_rows: Option<Result<Vec<(Row, NonZeroI64)>, String>>,
```

Comment:
We need this only because we can't put a batch of rows back into the `peek_iterator`, right? Could we just check for available space inside `rows_tx` before polling the `peek_iterator` and avoid the need for a stash? Looks like [try_reserve](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.try_reserve) lets you do that.